0029-loopback-transport/0029-loopback-transport.md

```diff
@@ -0,0 +1,50 @@
+
+## Chunked Flow
+
+Wnen sending multiple ILPv4 payments, as chunks in a chunked payment, the parameters passed from application layer to the transport layer are different. The application layer specifies a desired cummulative destination amount, and a deadline. The transport layer module then creates chunks of varying size and expiry duration, and accepts only the chunks whose exchange rate is best. If the cummulative value of chunks accepted so far is behind schedule for reaching the desired total before the deadline, chunks that arrive with slightly worse exchange rates will be accepted too. If the cummulative value is increasing ahead of schedule, the threshold will be adjusted to accept only the chunks with the very best exchange rates.
+
+## How to use
+
+When designing an application layer protocol, it's advisable to support both LT and PSK. Even though LT is far simpler than PSK, the extra loopback hop means that LT payments inherently have a higher latency than PSK payments, where the receiver can immediately fulfill incoming prepared payments. In situations where the sender can easily formulate instructions for the receiver about when to accept a payment and when not, and the sender trusts the receiver to execute these instructions in good faith, the speed up achieved with PSK means money needs to be tied up for less time, and this may translate to lower transaction fees, depending on connector pricing policies. So the application layer protocol needs to provide a choice between LT and PSK.
+
+If LT is chosen, the loopback link needs to be established; this will usually require some data exchange (e.g., one of the two parties sends connection details to the other party).
+
+The flow described above assumes one party only acts as the sender, and the other party only acts as the receiver, but once the loopback link is established, it can easily be used in both directions (in that case, each participant would send an IL-DCP request to the other participant, to ask for a loopback address).
+
+Once the loopback link is up, and each party who wants to act as a sender knows what their loopback address is, LT exposes two methods and one event to the application layer: one method for the one-shot flow, one for the chunked flow, one event that gets triggered whenever an incoming payment gets fulfilled, and one for when an outgoing payment gets fulfilled.
+
+It is up to the application layer protocol to decide how much to send, and whether to use a one-shot or a chunked flow.
+
+In future, hooks may be provided through which the application layer can also decide to reject incoming payments instead of looping them back to the appropriate sender.
+
+## Discovery
+
+A loopback server can be defined in the Loopback Server Definition (LSD) format, whose mime-type is `application/lsd+json`:
```

Comment:
Think this should be an extension of SPSP rather.
Example:

```json
{
  "destination_account": "example.ilpdemo.red.bob",
  "loopback": "wss+btp://token@example.com/",
  "maximum_destination_amount": "100000",
  "minimum_destination_amount": "10",
  "ledger_info": {
    "currency_code": "USD",
    "currency_scale": 2
  },
  "receiver_info": {
    "name": "Bob Dylan",
    "image_url": "https://red.ilpdemo.org/api/spsp/bob/profile_pic.jpg"
  }
}
```

I don't think it's necessary to put the `protocol` in as this should be determined from the loop back URI implicitly.

Reply:
The "destination_account" could be left out and established via ILDCP but having it in here is an optimization

Reply:
> Think this should be an extension of SPSP rather.

I could go either way with this, I kinda liked using the switch that we already have [at the payment-pointer level](https://github.com/interledger/rfcs/pull/413/files#diff-26073d3cbab98c2b76aeb396ea5ef212R103)

Reply:
I think it's too early to be fragmenting the system like that. SPSP v3 which supports more transport options seems better than a whole new setup protocol to me.

The ability to support other setup protocols via Payment Pointers is future proofing and loose coupling so that if someone has a better idea than SPSP we can still use Payment Pointers. 

That shouldn't be an excuse to not use SPSP when it actually already delivers everything we need for loopback IMO

Reply:
> it should be the client, not the server, who picks the transport

That is not how most protocol negotiation works. Servers usually provide options and clients pick from those. 

Payment Pointers is not a protocol for discovering a transport protocol, it's a protocol for discovering information about a receiver, including all the transport protocols it supports.

The protocol negotiation built into Payment Pointers is not for picking a transport protocol its for picking a setup protocol.

Unless loop-back is incompatible with SPSP I don't see why you'd invent a new setup protocol.

What prevents the SPSP response from providing multiple protocol options to the client?

```
{
  "destination_account": "example.ilpdemo.red.bob",
  "shared_secret": "vdytf65rdhgcioIYbuffGKItgRfg"
  "loopback": "wss+btp://token@example.com/",
  "balance": {
    "minimum": "0",
    "current": "-1000"
  },
  "asset_info": {
    "code": "USD",
    "scale": 2
  },
  "receiver_info": {
    "name": "Bob Dylan",
    "image_url": "https://red.ilpdemo.org/api/spsp/bob/profile_pic.jpg"
  }
}
```

Reply:
@adrianhopebailie you're making a valid point, so I'll split up this PR into one part which just describes the loopback transport (which I want to keep simple and finish quickly), and a second part (which can take longer and be a more fundamental cleanup) dealing with cleaning up our setup protocols so that they support multiple transports.